cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStoreTest.java

```diff
@@ -601,6 +608,106 @@ public void testGetAppClasses() throws Exception {
+        "retry", new PluginPropertyField("retries", "description", "int", false, false)
+      ), new HashSet<>(), ImmutableSet.of(Requirements.TRANSACTIONS, "spark2")
+    );
+
+    // required transaction, spark2 and some other requirement
+    PluginClass excludedPlugin4 = new PluginClass(
+      "A", "excludedPlugin4", "desc", "c.p1", "cfg",
+      ImmutableMap.of(
+        "threshold", new PluginPropertyField("thresh", "description", "double", true, false),
+        "retry", new PluginPropertyField("retries", "description", "int", false, false)
+      ), new HashSet<>(), ImmutableSet.of(Requirements.TRANSACTIONS, "spark2", "satisfied")
+    );
+
+    // required transaction and some other requirement
+    PluginClass excludedPlugin5 = new PluginClass(
+      "A", "excludedPlugin5", "desc", "c.p1", "cfg",
+      ImmutableMap.of(
+        "threshold", new PluginPropertyField("thresh", "description", "double", true, false),
+        "retry", new PluginPropertyField("retries", "description", "int", false, false)
+      ), new HashSet<>(), ImmutableSet.of(Requirements.TRANSACTIONS, "satisfied")
+    );
+
+    // requires some other requirement than transaction
+    PluginClass includedPlugin2 = new PluginClass(
+      "A", "includedPlugin2", "desc", "c.p1", "cfg",
+      ImmutableMap.of(
+        "threshold", new PluginPropertyField("thresh", "description", "double", true, false),
+        "retry", new PluginPropertyField("retries", "description", "int", false, false)
+      ), new HashSet<>(), ImmutableSet.of("noTransactionNeeded")
+    );
+
+    // requires multiple requirements other than the ones excluded
+    PluginClass includedPlugin3 = new PluginClass(
+      "A", "includedPlugin3", "desc", "c.p1", "cfg",
+      ImmutableMap.of(
+        "threshold", new PluginPropertyField("thresh", "description", "double", true, false),
+        "retry", new PluginPropertyField("retries", "description", "int", false, false)
+      ), new HashSet<>(), ImmutableSet.of("noTransactionNeeded", "spark1")
+    );
+
+    Id.Artifact artifactId = Id.Artifact.from(Id.Namespace.DEFAULT, "ArtifactWithTransactionalPlugins", "1.0.0");
+    ArtifactMeta artifactMeta = new ArtifactMeta(
+      ArtifactClasses.builder().addPlugins(includedPlugin1, includedPlugin2, includedPlugin3, excludedPlugin1,
+                                           excludedPlugin2, excludedPlugin3, excludedPlugin4, excludedPlugin5).build(),
+      ImmutableSet.of(parentArtifacts));
+    writeArtifact(artifactId, artifactMeta, "no-content");
+    ArtifactDescriptor artifactInfo = artifactStore.getArtifact(artifactId).getDescriptor();
+
+    // plugins which have transaction or spark as requirement should be excluded from plugins listed for the artifact
+    Map<ArtifactDescriptor, Set<PluginClass>> actual = artifactStore.getPluginClasses(NamespaceId.DEFAULT, artifactId);
+    ImmutableSet<PluginClass> expectedPlugins = ImmutableSet.of(includedPlugin1, includedPlugin2, includedPlugin3);
```

Comment:
Use `Set` instead of `ImmutableSet` as the type.

Reply:
Done.